instrumentation/grape/lib/opentelemetry/instrumentation/grape/handler.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Grape
+      # This class subscribes to the generated ActiveSupport notifications and generates spans based on them.
+      class Handler
+        SUBSCRIPTIONS = {
+          endpoint_run: 'endpoint_run.grape',
+          endpoint_render: 'endpoint_render.grape',
+          endpoint_run_filters: 'endpoint_run_filters.grape'
+        }.freeze
+
+        class << self
+          def subscribe
+            SUBSCRIPTIONS.each do |subscriber_method, event|
+              ::ActiveSupport::Notifications.subscribe(event) do |*args|
+                method(subscriber_method).call(*args)
+              end
+            end
+          end
+
+          private
+
+          def endpoint_run(name, start, finish, id, payload)
+            # TODO: check span.name and span.type
+            tracer.in_span(name, attributes: build_run_attributes(payload)) {}
+          end
+
+          def endpoint_render(name, start, finish, id, payload)
+            attributes = {
+              'component' => 'template',
+              'operation' => 'endpoint_render'
+            }
+            tracer.in_span(name, attributes: attributes) {}
+          end
+
+          def endpoint_run_filters(name, start, finish, id, payload)
+            filters = payload[:filters]
+            type = payload[:type]
+
+            # safe-guard to prevent submitting empty filters
+            # TODO: check if it's necessary
+            zero_length = (finish - start).zero?
+            return if (!filters || filters.empty?) || !type || zero_length
+
+            attributes = {
+              'component' => 'grape',
+              'operation' => 'endpoint_run_filters',
+              'grape.filter.type' => type.to_s
+            }
+            tracer.in_span(name, attributes: attributes) {}
+          end
+
+          def build_run_attributes(payload)
+            endpoint = payload.fetch(:endpoint)
+            request_method = endpoint.options.fetch(:method).first
+            path = endpoint_expand_path(endpoint)
+            api_instance = endpoint.options[:for]
+            {
+              'component' => 'web',
```

Comment:
Got this from here: https://github.com/catawiki/opentelemetry-ruby-contrib/blob/55b6be133bd038fc597d8a08f7da661734eb0d38/instrumentation/datadog-porting-guide.md#L100